In [15]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import cross_val_score
import numpy as np

## Part 1. Loading the dataset

In [16]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv('science_data_large.csv')
print(data.head(15))

print(data.info())
print(data.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [17]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X = data[['Temperature °C', 'Mols KCL']]  
y = data['Size nm^3']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [18]:
# Use sklearn to train a model on the training set
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = [[25, 0.1]]
predicted_size = linear_model.predict(sample_data)
print(f"Predicted slime size for temperature 25°C and Mols KCL 0.1: {predicted_size[0]} nm^3")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = linear_model.score(X_test, y_test)
print(f"Model R-squared score: {score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = linear_model.coef_
intercept = linear_model.intercept_

print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")


Predicted slime size for temperature 25°C and Mols KCL 0.1: -387634.54974246136 nm^3
Model R-squared score: 0.8552472077276096
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.479583408


/Users/<redacted>/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


The R-squared score of the model is 0.855. 
This score represents how well the linear regression model fits the data with a range from 0 to 1. 
An R-squared score of 0.855 means that 85.5% of the variance in slime size can be explained by the model based on temperature and mols of KCl. This suggests the model is a good fit. There is still about 14.5% of the variance that the model does not show.

Sample equation: $E = mc^2$

h(x) = -409391.48 + 866.15 * Temperature + 1032.70 * Mols KCl



## Part 4. Use Cross Validation

In [19]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

cv_scores = cross_val_score(linear_model, X, y, cv=5)

print(f"Cross-validation R-squared scores: {cv_scores}")

mean_cv_score = np.mean(cv_scores)
std_cv_score = np.std(cv_scores)

print(f"Mean R-squared score: {mean_cv_score}")
print(f"Standard deviation of R-squared score: {std_cv_score}")

# Report on their finding and their significance

Cross-validation R-squared scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean R-squared score: 0.8568167899144437
Standard deviation of R-squared score: 0.013466307372095953


The R-squared scores from the cross-validation runs were: [0.8392, 0.8705, 0.8587, 0.8720, 0.8436]
The mean R-squared score of 0.857 shows that, on average, the model explains around 85.7% of the variance in the slime growth based on temperature and mols of KCl. This means that the model fits the data pretty well. The standard deviation of 0.0135 indicates that the model's performance is consistent with different splits of the data as it generalizes well and is not too sensitive to the particular partitioning of the data. 

## Part 5. Using Polynomial Regression

In [20]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.

polynomial = PolynomialFeatures(degree=2)

X_poly = polynomial.fit_transform(X)

poly_model = LinearRegression()
poly_model.fit(X_poly, y)

poly_r2_score = poly_model.score(X_poly, y)
print(f"Polynomial Regression R-squared score: {poly_r2_score}")

poly_coefficients = poly_model.coef_
poly_intercept = poly_model.intercept_

print(f"Coefficients: {poly_coefficients}")
print(f"Intercept: {poly_intercept}")


Polynomial Regression R-squared score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23110281e-07 -1.05639941e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 1.6570091247558594e-05


h(x) = 0.0000166 + 12 * Temperature + 0 * Mols KCl 
       + 0 * (Temperature)^2 + 2 * (Mols KCl)^2 
       + 0.02857 * (Temperature * Mols KCl)